src/Ansatz/Chain.jl

```diff
@@ -547,6 +570,62 @@ function evolve_2site!(qtn::Chain, gate::Dense; threshold, maxdim)
     return qtn
 end
 
+"""
+    contract_θ(ψ::Chain, site::Site)
+
+For a [`Chain`](@ref) in the canonical form, returns the two-site wave function θ tensor
+(ΛᵢΓᵢΛᵢ₊₁Γᵢ₊₁Λᵢ₊₂), where i is the `site` index.
+"""
+function contract_θ(ψ::Chain, i::Site)
+    # TODO Check if ψ is in canonical form
+
+    0 < i.id < nsites(ψ) || throw(ArgumentError("Site index must be between 1 and $(nsites(ψ))"))
+
+    Λᵢ = i.id == 1 ? nothing : select(ψ, :between, Site(i.id - 1), i)
+    Λᵢ₊₂ = i.id == nsites(ψ) ? nothing : select(ψ, :between, i, Site(i.id + 1))
+
+    Γᵢ = select(ψ, :tensor, i)
+    Λᵢ₊₁ = select(ψ, :between, i, Site(i.id + 1))
+    Γᵢ₊₁ = select(ψ, :tensor, Site(i.id + 1))
+
+    tensors_to_contract = filter(x -> x !== nothing, [Λᵢ, Γᵢ, Λᵢ₊₁, Γᵢ₊₁, Λᵢ₊₂])
+    inds_to_contract = inds(Γᵢ) ∩ inds(Γᵢ₊₁)
+
+    return contract(tensors_to_contract...; dims = inds_to_contract)
+end
+
+"""
+    contract_θ!(ψ::Chain, site::Site)
+
+For a [`Chain`](@ref) in the canonical form, forms the two-site wave function θ with ΛᵢΓᵢΛᵢ₊₁Γᵢ₊₁Λᵢ₊₂,
```

Comment:
I think it would be less confusing if you use the notation $\Lambda_{(i-1, i)} \Gamma_i \Lambda_{(i,j)} \Gamma_j \Lambda_{(j,j+1)}$ instead.

Reply:
This for me is not ideal, I put the subindices there to explicitly show which $\Lambda$ and $\Gamma$ tensor we refer (we have a collection of $\Lambda^{[N-1]}$ and $\Gamma^{[N]}$ tensors). The indices you put are tensor indices, which is another thing. l ike the notation that [TenPy](https://arxiv.org/abs/1805.00055) uses, which is a superindex.